✨ Set up GitHub Copilot instructions for SSVC repository (#1012)
CERTCC/SSVC: .github/copilot-instructions.md

## Project Overview

SSVC is a modular decision-making framework for vulnerability management that includes:
- Python modules for decision points, decision tables, and outcomes
- MkDocs-based documentation website
- Interactive calculators and policy explorers
- JSON/CSV data files for decision tables
- Docker-based development and deployment

## Technology Stack

- **Primary Language**: Python 3.x
- **Package Management**: uv (package and project manager)
- **Documentation**: MkDocs with Material theme
- **Testing**: pytest
- **Data Models**: Pydantic for JSON schema validation
- **Scientific Computing**: NumPy, SciPy, scikit-learn
- **Web Framework**: FastAPI (for API endpoints)
- **Containerization**: Docker and Docker Compose

## Project Structure

- `/src/ssvc/` - Core Python modules for SSVC functionality
- `/docs/` - Markdown documentation source files
- `/data/` - JSON and CSV data files for decision tables
- `/test/` - Unit tests (located in `/src/test/`)
- `/docker/` - Docker configurations
- `/obsolete/` - Deprecated code (do not modify)

## Development Workflow

### Getting Started

1. **Set up development environment**: `make dev` or `uv sync --dev --project src`

> 2. **Run local documentation server**: `make docs_local` or `make docs` (Docker)

Reviewer: Suggest changing L39 and L40 to have commands of the same pattern: local, docker

Reviewer 2: Does this matter for copilot to know? These are not intended to be instructions for humans.

- Run markdown linting with auto-fix: `make mdlint_fix`
- Preview documentation changes: `make docs_local`
- Build documentation in Docker: `make docs`

### Common Commands

```bash
# Show available make targets
make help

# Set up dev environment
make dev

# Run tests locally
make test

# Run tests in Docker
make docker_test

# Serve documentation locally
make docs_local

# Regenerate JSON data files
make regenerate_json
```

## Coding Conventions

### Python Code

- Follow PEP 8 style guidelines
- Use type hints for function signatures
- Use Pydantic models for data validation
- Document classes and functions with docstrings
- Prefer explicit imports over wildcard imports
- Module structure uses absolute imports from `ssvc` package

### File Organization

- Decision points are in `/src/ssvc/decision_points/`
- Decision tables are in `/src/ssvc/decision_tables/`
- Tests mirror the source structure in `/src/test/`
- Documentation pages are in `/docs/`

### Naming Conventions

- Python files: `snake_case.py`
- Classes: `PascalCase`
- Functions/variables: `snake_case`
- Constants: `UPPER_SNAKE_CASE`

> - JSON/CSV files: `kebab-case.json`

Reviewer 2: I'd rather tell an LLM what we aspire to than what we've achieved so far.

## Testing Requirements

### Test Structure

- Unit tests use pytest framework
- Tests are located in `/src/test/`
- Test files follow pattern: `test_*.py`
- Run tests with: `make test` or `uv run --project src pytest -v`

### Test Coverage

- Write tests for new Python modules
- Ensure decision points and tables have corresponding tests
- Test JSON schema validation
- Validate data model serialization/deserialization

### Before Committing

1. Run all tests: `make test`
2. Ensure no test failures
3. Fix any linting issues: `make mdlint_fix`
4. Verify documentation builds: `make docs_local`

## Documentation

### Writing Documentation

- Documentation uses MkDocs with Material theme
- Files are in Markdown format in `/docs/`
- Use Python exec blocks for dynamic content generation
- Include examples and code snippets
- Follow existing documentation structure

### Documentation Features

- Automatic API documentation via mkdocstrings
- Code execution in Markdown via markdown-exec

> - BibTeX citations via mkdocs-bibtex

Reviewer 2: No what? We currently have that. We may not want it going forward. But then what do you change here?

> - BibTeX citations via mkdocs-bibtex

Reviewer: Are we using it anywhere? Is it relevant for copilot to know?

- Include markdown files with mkdocs-include-markdown-plugin

### Building Documentation

```bash
# Local preview (with hot reload)
make docs_local

# Docker build (production-like)
make docs

# Access at: http://localhost:8000/SSVC/
```

## Data Files

### JSON Files

- Located in `/data/json/`
- Generated from Python Pydantic models
- Use JSON schema validation
- Regenerate with: `make regenerate_json`

### CSV Files

- Located in `/data/csv/`
- Generated from Python modules
- Define decision table outcomes
- Primary way to customize SSVC for specific environments

## Common Pitfalls

1. **Import Paths**: Use absolute imports like `from ssvc.module import Class`, not relative imports
2. **PYTHONPATH**: When running scripts directly, set `export PYTHONPATH=$PYTHONPATH:$(pwd)/src`
3. **JSON Regeneration**: After modifying decision points/tables, regenerate JSON with `make regenerate_json`
4. **Docker Context**: Some make targets use Docker, others run locally - check the Makefile

> 5. **Package Management**: Use `uv` commands with `--project src` flag, not pip directly

Reviewer 2: @copilot PR #980 removed the need for `--project src` in uv commands because pyproject.toml was relocated to the top level directory of the repository.

- Unit tests use pytest framework
- Tests are located in `/src/test/`
- Test files follow pattern: `test_*.py`

> - Run tests with: `make test`

Reviewer 2: ```suggestion
- Run tests with: `make test` or `uv run pytest -v`
```

### Documentation Features

- Automatic API documentation via mkdocstrings
- Python module imports for dynamic content generation
- BibTeX citations via mkdocs-bibtex

> - Include markdown files by specifying them in mkdocs.yml

Reviewer 2: ```suggestion
- Add markdown files to site navigation by specifying them in `mkdocs.yml`
- Include markdown files in other markdown files with `mkdocs-include-markdown-plugin`
- Dynamically generate content from python code blocks using the `markdown-exec` plugin
```

- Located in `/data/json/`
- Generated from Python Pydantic models
- Use JSON schema validation

### CSV Files

- Located in `/data/csv/`
- Define decision table outcomes

> - Primary way to customize SSVC for specific environments

Reviewer 2: ```suggestion
- Generated from python modules (The python data objects are authoritative)
- Allows users to explore customizing SSVC for specific environments
```
